Add CMakeLists.txt.

devinamatthews opened the pull request:
This commit also fixes #2, adds pkgconfig support, as well as installs a cmake package config file when cmake is used.

Co-authored by: Christopher Hillenbrand <christopher.hillenbrand@yale.edu>
Co-authored by: Marco Barbone <mbarbone@flatironinstitute.org>

devinamatthews committed:
Add CMakeLists.txt.

This commit also fixes #2, adds pkgconfig support, as well as installs a cmake package config file when cmake is used.

Co-authored by: Christopher Hillenbrand <christopher.hillenbrand@yale.edu>
Co-authored by: Marco Barbone <...>

devinamatthews commented:
@chillenb @DiamonDinoia this PR is based on your versions of tci CMakeLists.txt with some modifications. Since I very new to cmake could you take a look and make sure I'm not doing something stupid? Also @diamondinoa would you like to provide a public or no-reply Github email so the commit can be properly attributed?

DiamonDinoia commented on CMakeLists.txt:
```diff
@@ -0,0 +1,454 @@
+set(ENABLE_BARRIER auto CACHE STRING "use a barrier of the specified type, one of auto, pthread_barrier, \
+spin_barrier [default=auto]")
+
+set(ENABLE_THREAD_MODEL "" CACHE STRING "enable threading with the specified model, one of none, auto, task, \
```
default is auto but the string is empty here.

DiamonDinoia commented on CMakeLists.txt:
```diff
@@ -0,0 +1,454 @@
+set(ENABLE_THREAD_MODEL "" CACHE STRING "enable threading with the specified model, one of none, auto, task, \
+openmp, pthreads, tbb, omptask, winthreads, ppl, or dispatch [default=auto]")
+
+set(ENABLE_THREADING "" CACHE STRING "synonym for ENABLE_THREAD_MODEL")
```
we should specify the default here too. this can be an option()

devinamatthews replied:
My intent was to mirror what configure.ac does, which is use --enable-thread-model if supplied, and --enable-threading as a backup. The latter is the option in TBLIS so it will get inherited when TCI is used there.

I checked that the used value does come out to auto, but you are correct the doc string is a bit of a mismatch.

devinamatthews replied:
I think I can rework the logic so that both variables can default to `auto` but still overrride correctly. Essentially `ENABLE_THREADING != auto && ENABLE_THREAD_MODEL == auto` will use ENABLE_THREADING.

DiamonDinoia commented on CMakeLists.txt:
```diff
@@ -0,0 +1,454 @@
+endif()
+
+# Default to disabled for all mutex/barrier/thread options
+set(USE_PTHREAD_MUTEX 0)
+set(USE_PTHREAD_SPINLOCK 0)
+set(USE_OSX_SPINLOCK 0)
+set(USE_OMP_LOCK 0)
+set(USE_ATOMIC_SPINLOCK 0)
+
+set(USE_PTHREAD_BARRIER 0)
+set(USE_SPIN_BARRIER 0)
+
+set(USE_PTHREADS_THREADS 0)
+set(USE_OPENMP_THREADS 0)
+set(USE_TBB_THREADS 0)
+set(USE_PPL_THREADS 0)
+set(USE_OMPTASK_THREADS 0)
+set(USE_DISPATCH_THREADS 0)
+set(USE_WINDOWS_THREADS 0)
```
these should be all option: https://cmake.org/cmake/help/latest/command/option.html

devinamatthews replied:
I prefer to keep these internal and use ENABLE_THREAD_MODEL, ENABLE_MUTEX, and ENABLE_BARRIER.

chillenb replied:
I think it might be ok not to use `option` for these because user input will get overridden; these are set later on depending on `ENABLE_THREAD_MODEL`, etc.

DiamonDinoia commented on CMakeLists.txt:
```diff
@@ -0,0 +1,454 @@
+    LANGUAGES C CXX
+)
+
+set(CMAKE_CXX_STANDARD 20)
```
I do not set this globally usually unless I am prototyping.
set_target_properties( target PROPERTIES
  CXX_STANDARD 20
  CXX_STANDARD_REQUIRED ON
  CXX_EXTENSIONS OFF
)

devinamatthews replied:
Thx will change.

DiamonDinoia commented on CMakeLists.txt:
```diff
@@ -0,0 +1,454 @@
+
+# Detect pthreads
+if(CMAKE_USE_PTHREADS_INIT)
+    set(HAVE_PTHREADS_THREADS 1)
```
Instead of 1 I think the standard is to use ON

DiamonDinoia commented on CMakeLists.txt:
```diff
@@ -0,0 +1,454 @@
+endif()
+message(STATUS "mutex type selected: ${ENABLE_MUTEX}")
+
+if(ENABLE_MUTEX STREQUAL pthread_mutex)
+    if(NOT HAVE_PTHREAD_MUTEX)
+        message(FATAL_ERROR "pthread mutex requested but not available")
+    endif()
+    set(USE_PTHREAD_MUTEX 1)
+elseif(ENABLE_MUTEX STREQUAL pthread_spinlock)
+    if(NOT HAVE_PTHREAD_SPINLOCK)
+        message(FATAL_ERROR "pthread spinlock requested but not available")
+    endif()
+    set(USE_PTHREAD_SPINLOCK 1)
+elseif(ENABLE_MUTEX STREQUAL os_unfair_lock)
+    if(NOT HAVE_OS_UNFAIR_LOCK)
+        message(FATAL_ERROR "macOS os_unfair_lock requested but not available")
+    endif()
+    set(USE_OS_UNFAIR_LOCK 1)
+elseif(ENABLE_MUTEX STREQUAL osx_spinlock)
+    if(NOT HAVE_OSX_SPINLOCK)
+        message(FATAL_ERROR "OSX OSSpinLock requested but not available")
+    endif()
+    set(USE_OSX_SPINLOCK 1)
+elseif (ENABLE_MUTEX STREQUAL omp_lock)
+    if(NOT HAVE_OMP_LOCK)
+        message(FATAL_ERROR "omp_lock requested but not available")
+    endif()
+    set(USE_OMP_LOCK 1)
+elseif (ENABLE_MUTEX STREQUAL atomic_spinlock)
+    set(USE_ATOMIC_SPINLOCK 1)
+else()
+    message(FATAL_ERROR "invalid mutex type specified")
+endif()
```
It might be less error prone doing a loop over tuples:

```cmake
# Function to handle a generic choice from a list of 4-tuples.
# Each tuple: (choice_key, feature_check_var, variable_to_set, error_message)
function(handle_choice choice_list selected_key)
    list(LENGTH choice_list len)
    math(EXPR n "${len} / 4")

    set(found FALSE)

    foreach(i RANGE 0 ${n}-1)
        math(EXPR idx0 "4 * ${i}")
        math(EXPR idx1 "${idx0} + 1")
        math(EXPR idx2 "${idx0} + 2")
        math(EXPR idx3 "${idx0} + 3")

        list(GET choice_list ${idx0} choice_key)
        list(GET choice_list ${idx1} feature_var)
        list(GET choice_list ${idx2} set_var)
        list(GET choice_list ${idx3} error_msg)

        if("${selected_key}" STREQUAL "${choice_key}")
            set(found TRUE)

            if(NOT "${feature_var}" STREQUAL "")
                if(NOT DEFINED ${feature_var} OR NOT ${${feature_var}})
                    message(FATAL_ERROR "${error_msg}")
                endif()
            endif()

            set(${set_var} 1 PARENT_SCOPE)
            return()
        endif()
    endforeach()

    if(NOT found)
        message(FATAL_ERROR "Invalid choice: '${selected_key}'")
    endif()
endfunction()


set(MUTEX_CHOICES
    pthread_mutex     HAVE_PTHREAD_MUTEX     USE_PTHREAD_MUTEX     "pthread mutex requested but not available"
    pthread_spinlock  HAVE_PTHREAD_SPINLOCK  USE_PTHREAD_SPINLOCK  "pthread spinlock requested but not available"
    os_unfair_lock    HAVE_OS_UNFAIR_LOCK    USE_OS_UNFAIR_LOCK    "macOS os_unfair_lock requested but not available"
    osx_spinlock      HAVE_OSX_SPINLOCK      USE_OSX_SPINLOCK      "OSX OSSpinLock requested but not available"
    omp_lock          HAVE_OMP_LOCK          USE_OMP_LOCK          "omp_lock requested but not available"
    atomic_spinlock   ""                     USE_ATOMIC_SPINLOCK   ""
)

set(ENABLE_MUTEX "pthread_mutex")
handle_choice("${MUTEX_CHOICES}" "${ENABLE_MUTEX}")

```

DiamonDinoia commented on CMakeLists.txt:
```diff
@@ -0,0 +1,454 @@
+endif()
+
+if(USE_PTHREADS_THREADS OR USE_PTHREAD_MUTEX OR USE_PTHREAD_BARRIER OR USE_PTHREAD_SPINLOCK)
+    list(APPEND EXTRA_LIBS Threads::Threads)
```
Did we check that we have threads here?
Something feels off from the test above and this.

Also find package might already do some of the checks: https://cmake.org/cmake/help/latest/module/FindThreads.html

chillenb replied:
There's a `find_package(Threads)` at the top, which does the check and only sets `CMAKE_USE_PTHREADS_INIT` if the thread library is pthread-compatible. This was my doing; it was intended to be a fairly close translation of the original configure.ac logic, but as a result it definitely feels un-CMake-like.

devinamatthews replied:
The output of `find_package(Threads)` is checked for `HAVE_PTHREADS_THREADS` which guards `USE_PTHREADS_THREADS` (and similarly for pthread_mutex, pthread_barrier, and pthead_spinlock).

chillenb commented:
Since the public header files don't use Threads and OpenMP (and other EXTRA_LIBS) directly, I think they can be made `PRIVATE`. (Otherwise, I found during testing that TCIConfig.cmake would have to run `find_dependency` for each of them.) Linking these `PUBLIC` was my original suggestion, oops.

There's also a `sqrt` and `lrint` somewhere that require the math library.

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6a622b3..f515e57 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -12,7 +12,7 @@ project(TCI
 
 set(CMAKE_CXX_STANDARD 20)
 
-include(CheckIncludefile)
+include(CheckIncludeFile)
 include(CheckSymbolExists)
 include(CheckCSourceCompiles)
 include(CMakePackageConfigHelpers)
@@ -385,7 +385,12 @@ tci/slot.c
 tci/work_item.c
 )
 
-target_link_libraries(tci PUBLIC ${EXTRA_LIBS})
+find_library(TCI_MATH_LIBRARY m)
+if(TCI_MATH_LIBRARY)
+    target_link_libraries(tci PRIVATE ${TCI_MATH_LIBRARY})
+endif()
+
+target_link_libraries(tci PRIVATE ${EXTRA_LIBS})
 
 set_target_properties(tci PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/lib)
 
```

devinamatthews commented:
Re OpenMP/Threads, if a static lib is built, then marking these PUBLIC will make sure downstream projects link the right libraries, yes?

devinamatthews commented:
Re libm, I prefer to link it only when absolutely necessary, since e.g. the Intel compilers have their own implementation and -lm messes things up. Perhaps would be better to check for those symbols without -lm and only link if not found.

devinamatthews committed:
Fix usage of CMAKE_CXX_STANDARD...

...and make `auto` the default for both ENABLE_THREAD_MODEL and
ENABLE_THREADING.

devinamatthews committed:
Typo

chillenb commented:
> Re libm, I prefer to link it only when absolutely necessary, since e.g. the Intel compilers have their own implementation and -lm messes things up. Perhaps would be better to check for those symbols without -lm and only link if not found.

That seems like a frustrating issue. Maybe a conditional link like
```cmake
if(CMAKE_C_COMPILER_ID STREQUAL "Clang" OR CMAKE_C_COMPILER_ID STREQUAL "GNU")
  target_link_libraries(tci PRIVATE m)
endif()
```
could be helpful, if it doesn't disgust you too much XD

> Re OpenMP/Threads, if a static lib is built, then marking these PUBLIC will make sure downstream projects link the right libraries, yes?

It turns out CMake handles this case correctly. In the words of CMake guru Craig Scott (Professional CMake):

> 23.2. Linking Static Libraries 
As first mentioned back in Section 16.2.1, “Linking Libraries”, CMake handles some special cases specific to linking static libraries. If a library A is listed as a PRIVATE dependency for a static library target B, then A will effectively be treated as a PUBLIC dependency as far as linking is concerned (and only for linking). This is because the private A library will still need to be added to the linker command line of anything linking to B in order for symbols from A to be found at link time. If B was a shared library, the private library A that it depends on would not need to be listed on the linker command line. CMake handles all this transparently, so the developer typically doesn’t need to concern themselves with the details beyond specifying the PUBLIC, PRIVATE, and INTERFACE  dependencies with target_link_libraries(). 

> For the curious, when exporting the A and B targets (discussed in Section 35.3, “Installing Exports”), CMake will use the `$<LINK_ONLY:...>` generator expression to implement the "only for linking" logic. Projects won’t typically need to use `$<LINK_ONLY:...>` directly themselves, but its use can be seen in the export files that CMake generates.

So if Threads and OpenMP are linked `PRIVATE` and TCI is built as a **static** library, downstream projects will link Threads and OpenMP as they should!!! (I checked this). But they will need to find Threads and OpenMP somehow, either by doing
```cmake
find_package(Threads)
find_package(OpenMP)
find_package(TCI)
```

or with the following in TCIConfig.cmake
```cmake
include(CMakeFindDependencyMacro)
find_dependency(Threads QUIET)
find_dependency(OpenMP QUIET)
include(${CMAKE_CURRENT_LIST_DIR}/TCITargets.cmake)
```

chillenb commented:
🫡
Thanks for suffering through CMake for the sake of the community.
I will be happy to help with anything that comes up!

DiamonDinoia commented:
```cmake
    set_source_files_properties(tci/communicator.c PROPERTIES LANGUAGE CXX)
    set_source_files_properties(tci/task_set.c PROPERTIES LANGUAGE CXX)
```

Is there any reason this has to be a .c instead of a .cpp? 
I would say this breaks some IDEs that relies on cmake.

devinamatthews committed:
Several changes:

- Add static build option.
- Rearrange source and include files to satisfy CMake's dislike
  for top-level include files.
- Separate C and C++ code in both source files and headers.

devinamatthews commented:
I think the last commit fixes all of the outstanding issues, and makes some long-needed improvements (like better separating the C and C++ bits). If there are no other major issues I'll merge this and move on the other TBLIS dependencies (should be much easier) and then TBLIS proper.

DiamonDinoia commented:
> I think the last commit fixes all of the outstanding issues, and makes some long-needed improvements (like better separating the C and C++ bits). If there are no other major issues I'll merge this and move on the other TBLIS dependencies (should be much easier) and then TBLIS proper.

I agree. This looks good.

I would only do:
```cmake
set_target_properties(tci-static PROPERTIES OUTPUT_NAME "tci")
```

as is the output artifacts will be tci.so and tci-static.a. WIth the line above they will be tci.so and tci.a which is more consistent.
